File: microgpt.ts
/**
 * The most atomic way to train and run inference for a GPT in pure, dependency-free TypeScript.
 * This file is the complete algorithm.
 * Everything else is just efficiency.
 *
 * Translated from @karpathy's microgpt.py
 * Run: npx tsx microgpt.ts
 */

import { readFileSync, writeFileSync, existsSync } from "fs";
import { get } from "https";

// ============================================================================
// 伪随机数生成器（Mulberry32），替代 Python 的 random 模块
// 使用固定种子 42，保证每次运行结果可复现
// ============================================================================
class RNG {
  private state: number;
  constructor(seed: number) {
    this.state = seed;
  }
  /** 返回 [0, 1) 的均匀分布随机数 */
  random(): number {
    this.state |= 0;
    this.state = (this.state + 0x6d2b79f5) | 0;
    let t = Math.imul(this.state ^ (this.state >>> 15), 1 | this.state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }
  /** Box-Muller 变换：生成标准正态分布随机数 */
  gauss(mean: number = 0, std: number = 1): number {
    const u1 = this.random();
    const u2 = this.random();
    return mean + std * Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  }
  /** Fisher-Yates 洗牌算法 */
  shuffle<T>(arr: T[]): void {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Math.floor(this.random() * (i + 1));
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }
}

const rng = new RNG(42);

// ============================================================================
// 自动微分引擎（Autograd）
// Value 类包装标量值并记录计算图，backward() 通过链式法则自动计算梯度
// ============================================================================
class Value {
  data: number;
  grad: number = 0;
  private children: Value[];
  private localGrads: number[];

  constructor(data: number, children: Value[] = [], localGrads: number[] = []) {
    this.data = data;
    this.children = children;
    this.localGrads = localGrads;
  }

  add(other: Value): Value {
    return new Value(this.data + other.data, [this, other], [1, 1]);
  }
  addS(s: number): Value {
    return this.add(new Value(s));
  }
  mul(other: Value): Value {
    return new Value(this.data * other.data, [this, other], [other.data, this.data]);
  }
  mulS(s: number): Value {
    return this.mul(new Value(s));
  }
  pow(n: number): Value {
    return new Value(this.data ** n, [this], [n * this.data ** (n - 1)]);
  }
  log(): Value {
    return new Value(Math.log(this.data), [this], [1 / this.data]);
  }
  exp(): Value {
    const e = Math.exp(this.data);
    return new Value(e, [this], [e]);
  }
  relu(): Value {
    const d = this.data > 0 ? this.data : 0;
    const g = this.data > 0 ? 1 : 0;
    return new Value(d, [this], [g]);
  }
  neg(): Value {
    return this.mulS(-1);
  }
  sub(other: Value): Value {
    return this.add(other.neg());
  }
  div(other: Value): Value {
    return this.mul(other.pow(-1));
  }

  backward(): void {
    const topo: Value[] = [];
    const visited = new Set<Value>();
    const buildTopo = (v: Value) => {
      if (visited.has(v)) return;
      visited.add(v);
      for (const child of v.children) buildTopo(child);
      topo.push(v);
    };
    buildTopo(this);
    this.grad = 1;
    for (let i = topo.length - 1; i >= 0; i--) {
      const v = topo[i];
      for (let j = 0; j < v.children.length; j++) {
        v.children[j].grad += v.localGrads[j] * v.grad;
      }
    }
  }
}

// ============================================================================
// 辅助函数
// ============================================================================
function sumValues(vals: Value[]): Value {
  let s = new Value(0);
  for (const v of vals) s = s.add(v);
  return s;
}

function linear(x: Value[], w: Value[][]): Value[] {
  return w.map((wo) => {
    let s = new Value(0);
    for (let j = 0; j < x.length; j++) s = s.add(wo[j].mul(x[j]));
    return s;
  });
}

function softmax(logits: Value[]): Value[] {
  let maxVal = logits[0].data;
  for (let i = 1; i < logits.length; i++) {
    if (logits[i].data > maxVal) maxVal = logits[i].data;
  }
  const exps = logits.map((v) => v.addS(-maxVal).exp());
  const total = sumValues(exps);
  return exps.map((e) => e.div(total));
}

function rmsnorm(x: Value[]): Value[] {
  let s = new Value(0);
  for (const xi of x) s = s.add(xi.mul(xi));
  const ms = s.mulS(1 / x.length);
  const scale = ms.addS(1e-5).pow(-0.5);
  return x.map((xi) => xi.mul(scale));
}

// ============================================================================
// 下载数据集 & 启动训练
// ============================================================================
function downloadFile(url: string, dest: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const file = require("fs").createWriteStream(dest);
    get(url, (resp) => {
      // handle redirects
      if (resp.statusCode === 301 || resp.statusCode === 302) {
        file.close();
        downloadFile(resp.headers.location!, dest).then(resolve).catch(reject);
        return;
      }
      resp.pipe(file);
      file.on("finish", () => { file.close(); resolve(); });
    }).on("error", reject);
  });
}

async function main() {
  // --- 数据集加载 ---
  if (!existsSync("input.txt")) {
    console.log("Downloading input.txt...");
    await downloadFile(
      "https://raw.githubusercontent.com/karpathy/makemore/988aa59/names.txt",
      "input.txt"
    );
  }
  const docs = readFileSync("input.txt", "utf-8")
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l.length > 0);
  rng.shuffle(docs);
  console.log(`num docs: ${docs.length}`);

  // --- Tokenizer ---
  const charSet = new Set<string>();
  for (const doc of docs) for (const ch of doc) charSet.add(ch);
  const uchars = [...charSet].sort();
  const BOS = uchars.length;
  const vocabSize = uchars.length + 1;
  console.log(`vocab size: ${vocabSize}`);

  const charIndex = new Map<string, number>();
  uchars.forEach((ch, i) => charIndex.set(ch, i));

  // --- 模型参数初始化 ---
  const nLayer = 1;
  const nEmbd = 16;
  const blockSize = 16;
  const nHead = 4;
  const headDim = nEmbd / nHead;

  const matrix = (nout: number, nin: number, std = 0.08): Value[][] => {
    const m: Value[][] = [];
    for (let i = 0; i < nout; i++) {
      const row: Value[] = [];
      for (let j = 0; j < nin; j++) row.push(new Value(rng.gauss(0, std)));
      m.push(row);
    }
    return m;
  };

  const stateDict: Record<string, Value[][]> = {
    wte: matrix(vocabSize, nEmbd),
    wpe: matrix(blockSize, nEmbd),
    lm_head: matrix(vocabSize, nEmbd),
  };
  for (let i = 0; i < nLayer; i++) {
    stateDict[`layer${i}.attn_wq`] = matrix(nEmbd, nEmbd);
    stateDict[`layer${i}.attn_wk`] = matrix(nEmbd, nEmbd);
    stateDict[`layer${i}.attn_wv`] = matrix(nEmbd, nEmbd);
    stateDict[`layer${i}.attn_wo`] = matrix(nEmbd, nEmbd);
    stateDict[`layer${i}.mlp_fc1`] = matrix(4 * nEmbd, nEmbd);
    stateDict[`layer${i}.mlp_fc2`] = matrix(nEmbd, 4 * nEmbd);
  }

  // 保持与 Python dict 插入顺序一致
  const paramKeys = ["wte", "wpe", "lm_head"];
  for (let i = 0; i < nLayer; i++) {
    paramKeys.push(
      `layer${i}.attn_wq`, `layer${i}.attn_wk`, `layer${i}.attn_wv`,
      `layer${i}.attn_wo`, `layer${i}.mlp_fc1`, `layer${i}.mlp_fc2`
    );
  }
  const params: Value[] = [];
  for (const key of paramKeys) {
    for (const row of stateDict[key]) params.push(...row);
  }
  console.log(`num params: ${params.length}`);

  // --- GPT 前向传播 ---
  const gpt = (tokenId: number, posId: number, keys: Value[][][], values: Value[][][]): Value[] => {
    const tokEmb = stateDict["wte"][tokenId];
    const posEmb = stateDict["wpe"][posId];
    let x = tokEmb.map((t, i) => t.add(posEmb[i]));
    x = rmsnorm(x);

    for (let li = 0; li < nLayer; li++) {
      // 1) Multi-head Attention
      const xResidual1 = x;
      x = rmsnorm(x);
      const q = linear(x, stateDict[`layer${li}.attn_wq`]);
      const k = linear(x, stateDict[`layer${li}.attn_wk`]);
      const v = linear(x, stateDict[`layer${li}.attn_wv`]);
      keys[li].push(k);
      values[li].push(v);
      const xAttn: Value[] = [];
      for (let h = 0; h < nHead; h++) {
        const hs = h * headDim;
        const qH = q.slice(hs, hs + headDim);
        const kH = keys[li].map((ki) => ki.slice(hs, hs + headDim));
        const vH = values[li].map((vi) => vi.slice(hs, hs + headDim));
        const scale = 1 / Math.sqrt(headDim);
        const attnLogits: Value[] = [];
        for (let t = 0; t < kH.length; t++) {
          let s = new Value(0);
          for (let j = 0; j < headDim; j++) s = s.add(qH[j].mul(kH[t][j]));
          attnLogits.push(s.mulS(scale));
        }
        const attnWeights = softmax(attnLogits);
        for (let j = 0; j < headDim; j++) {
          let s = new Value(0);
          for (let t = 0; t < vH.length; t++) s = s.add(attnWeights[t].mul(vH[t][j]));
          xAttn.push(s);
        }
      }
      x = linear(xAttn, stateDict[`layer${li}.attn_wo`]);
      x = x.map((a, i) => a.add(xResidual1[i]));

      // 2) MLP
      const xResidual2 = x;
      x = rmsnorm(x);
      x = linear(x, stateDict[`layer${li}.mlp_fc1`]);
      x = x.map((xi) => xi.relu());
      x = linear(x, stateDict[`layer${li}.mlp_fc2`]);
      x = x.map((a, i) => a.add(xResidual2[i]));
    }

    return linear(x, stateDict["lm_head"]);
  };

  // --- Adam 优化器 ---
  const learningRate = 0.01, beta1 = 0.85, beta2 = 0.99, epsAdam = 1e-8;
  const mBuf = new Float64Array(params.length);
  const vBuf = new Float64Array(params.length);

  // --- 训练循环 ---
  const numSteps = 1000;
  for (let step = 0; step < numSteps; step++) {
    const doc = docs[step % docs.length];
    const tokens = [BOS, ...Array.from(doc).map((ch) => charIndex.get(ch)!), BOS];
    const n = Math.min(blockSize, tokens.length - 1);

    const keys: Value[][][] = Array.from({ length: nLayer }, () => []);
    const vals: Value[][][] = Array.from({ length: nLayer }, () => []);
    const losses: Value[] = [];
    for (let posId = 0; posId < n; posId++) {
      const tokenId = tokens[posId];
      const targetId = tokens[posId + 1];
      const logits = gpt(tokenId, posId, keys, vals);
      const probs = softmax(logits);
      losses.push(probs[targetId].log().neg());
    }
    const loss = sumValues(losses).mulS(1 / n);
    loss.backward();

    const lrT = learningRate * (1 - step / numSteps);
    const stepP1 = step + 1;
    for (let i = 0; i < params.length; i++) {
      const p = params[i];
      mBuf[i] = beta1 * mBuf[i] + (1 - beta1) * p.grad;
      vBuf[i] = beta2 * vBuf[i] + (1 - beta2) * p.grad * p.grad;
      const mHat = mBuf[i] / (1 - beta1 ** stepP1);
      const vHat = vBuf[i] / (1 - beta2 ** stepP1);
      p.data -= lrT * mHat / (Math.sqrt(vHat) + epsAdam);
      p.grad = 0;
    }

    process.stdout.write(`step ${String(step + 1).padStart(4)} / ${numSteps} | loss ${loss.data.toFixed(4)}\r`);
  }

  // --- 推理 ---
  const temperature = 0.5;
  console.log("\n--- inference (new, hallucinated names) ---");
  for (let sampleIdx = 0; sampleIdx < 20; sampleIdx++) {
    const keys: Value[][][] = Array.from({ length: nLayer }, () => []);
    const vals: Value[][][] = Array.from({ length: nLayer }, () => []);
    let tokenId = BOS;
    const sample: string[] = [];
    for (let posId = 0; posId < blockSize; posId++) {
      const logits = gpt(tokenId, posId, keys, vals);
      const probs = softmax(logits.map((l) => l.mulS(1 / temperature)));
      // 加权随机采样
      const r = rng.random();
      let cum = 0;
      tokenId = 0;
      for (let i = 0; i < probs.length; i++) {
        cum += probs[i].data;
        if (r <= cum) { tokenId = i; break; }
      }
      if (tokenId === BOS) break;
      sample.push(uchars[tokenId]);
    }
    console.log(`sample ${String(sampleIdx + 1).padStart(2)}: ${sample.join("")}`);
  }
}

main();

File: microgpt.go
// The most atomic way to train and run inference for a GPT in pure, dependency-free Go.
// This file is the complete algorithm.
// Everything else is just efficiency.
//
// Translated from @karpathy's microgpt.py

package main

import (
	"fmt"
	"io"
	"math"
	"math/rand"
	"net/http"
	"os"
	"sort"
	"strings"
)

// Let there be Autograd to recursively apply the chain rule through a computation graph
type Value struct {
	Data       float64
	Grad       float64
	children   []*Value
	localGrads []float64
}

func NewValue(data float64) *Value {
	return &Value{Data: data}
}

func NewValueWithChildren(data float64, children []*Value, localGrads []float64) *Value {
	return &Value{Data: data, children: children, localGrads: localGrads}
}

func (a *Value) Add(b *Value) *Value {
	return NewValueWithChildren(a.Data+b.Data, []*Value{a, b}, []float64{1, 1})
}

func (a *Value) AddScalar(s float64) *Value {
	return a.Add(NewValue(s))
}

func (a *Value) Mul(b *Value) *Value {
	return NewValueWithChildren(a.Data*b.Data, []*Value{a, b}, []float64{b.Data, a.Data})
}

func (a *Value) MulScalar(s float64) *Value {
	return a.Mul(NewValue(s))
}

func (a *Value) Pow(other float64) *Value {
	return NewValueWithChildren(math.Pow(a.Data, other), []*Value{a}, []float64{other * math.Pow(a.Data, other-1)})
}

func (a *Value) Log() *Value {
	return NewValueWithChildren(math.Log(a.Data), []*Value{a}, []float64{1.0 / a.Data})
}

func (a *Value) Exp() *Value {
	e := math.Exp(a.Data)
	return NewValueWithChildren(e, []*Value{a}, []float64{e})
}

func (a *Value) ReLU() *Value {
	d := 0.0
	g := 0.0
	if a.Data > 0 {
		d = a.Data
		g = 1.0
	}
	return NewValueWithChildren(d, []*Value{a}, []float64{g})
}

func (a *Value) Neg() *Value {
	return a.MulScalar(-1)
}

func (a *Value) Sub(b *Value) *Value {
	return a.Add(b.Neg())
}

func (a *Value) Div(b *Value) *Value {
	return a.Mul(b.Pow(-1))
}

func (a *Value) Backward() {
	// Topological sort
	topo := make([]*Value, 0)
	visited := make(map[*Value]bool)
	var buildTopo func(v *Value)
	buildTopo = func(v *Value) {
		if visited[v] {
			return
		}
		visited[v] = true
		for _, child := range v.children {
			buildTopo(child)
		}
		topo = append(topo, v)
	}
	buildTopo(a)
	a.Grad = 1
	for i := len(topo) - 1; i >= 0; i-- {
		v := topo[i]
		for j, child := range v.children {
			child.Grad += v.localGrads[j] * v.Grad
		}
	}
}

// sum of []*Value -> *Value
func sumValues(vals []*Value) *Value {
	s := NewValue(0)
	for _, v := range vals {
		s = s.Add(v)
	}
	return s
}

// Define the model architecture functions
func linear(x []*Value, w [][]*Value) []*Value {
	out := make([]*Value, len(w))
	for i, wo := range w {
		s := NewValue(0)
		for j, xi := range x {
			s = s.Add(wo[j].Mul(xi))
		}
		out[i] = s
	}
	return out
}

func softmax(logits []*Value) []*Value {
	maxVal := logits[0].Data
	for _, v := range logits[1:] {
		if v.Data > maxVal {
			maxVal = v.Data
		}
	}
	exps := make([]*Value, len(logits))
	for i, v := range logits {
		exps[i] = v.AddScalar(-maxVal).Exp()
	}
	total := sumValues(exps)
	out := make([]*Value, len(logits))
	for i, e := range exps {
		out[i] = e.Div(total)
	}
	return out
}

func rmsnorm(x []*Value) []*Value {
	n := len(x)
	s := NewValue(0)
	for _, xi := range x {
		s = s.Add(xi.Mul(xi))
	}
	ms := s.MulScalar(1.0 / float64(n))
	scale := ms.AddScalar(1e-5).Pow(-0.5)
	out := make([]*Value, n)
	for i, xi := range x {
		out[i] = xi.Mul(scale)
	}
	return out
}

func main() {
	// Let there be order among chaos
	rng := rand.New(rand.NewSource(42))

	// Let there be a Dataset
	if _, err := os.Stat("input.txt"); os.IsNotExist(err) {
		fmt.Println("Downloading input.txt...")
		resp, err := http.Get("https://raw.githubusercontent.com/karpathy/makemore/988aa59/names.txt")
		if err != nil {
			panic(err)
		}
		defer resp.Body.Close()
		data, _ := io.ReadAll(resp.Body)
		os.WriteFile("input.txt", data, 0644)
	}
	fileData, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	lines := strings.Split(string(fileData), "\n")
	docs := make([]string, 0, len(lines))
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			docs = append(docs, line)
		}
	}
	rng.Shuffle(len(docs), func(i, j int) { docs[i], docs[j] = docs[j], docs[i] })
	fmt.Printf("num docs: %d\n", len(docs))

	// Let there be a Tokenizer
	charSet := make(map[rune]bool)
	for _, doc := range docs {
		for _, ch := range doc {
			charSet[ch] = true
		}
	}
	uchars := make([]rune, 0, len(charSet))
	for ch := range charSet {
		uchars = append(uchars, ch)
	}
	sort.Slice(uchars, func(i, j int) bool { return uchars[i] < uchars[j] })
	BOS := len(uchars)
	vocabSize := len(uchars) + 1
	fmt.Printf("vocab size: %d\n", vocabSize)

	// charIndex for fast lookup
	charIndex := make(map[rune]int)
	for i, ch := range uchars {
		charIndex[ch] = i
	}

	// Initialize the parameters
	nLayer := 1
	nEmbd := 16
	blockSize := 16
	nHead := 4
	headDim := nEmbd / nHead

	matrix := func(nout, nin int) [][]*Value {
		std := 0.08
		m := make([][]*Value, nout)
		for i := range m {
			m[i] = make([]*Value, nin)
			for j := range m[i] {
				m[i][j] = NewValue(rng.NormFloat64() * std)
			}
		}
		return m
	}

	stateDict := make(map[string][][]*Value)
	stateDict["wte"] = matrix(vocabSize, nEmbd)
	stateDict["wpe"] = matrix(blockSize, nEmbd)
	stateDict["lm_head"] = matrix(vocabSize, nEmbd)
	for i := 0; i < nLayer; i++ {
		stateDict[fmt.Sprintf("layer%d.attn_wq", i)] = matrix(nEmbd, nEmbd)
		stateDict[fmt.Sprintf("layer%d.attn_wk", i)] = matrix(nEmbd, nEmbd)
		stateDict[fmt.Sprintf("layer%d.attn_wv", i)] = matrix(nEmbd, nEmbd)
		stateDict[fmt.Sprintf("layer%d.attn_wo", i)] = matrix(nEmbd, nEmbd)
		stateDict[fmt.Sprintf("layer%d.mlp_fc1", i)] = matrix(4*nEmbd, nEmbd)
		stateDict[fmt.Sprintf("layer%d.mlp_fc2", i)] = matrix(nEmbd, 4*nEmbd)
	}

	// Flatten params — iterate in a deterministic order matching Python's dict insertion order
	paramKeys := []string{"wte", "wpe", "lm_head"}
	for i := 0; i < nLayer; i++ {
		paramKeys = append(paramKeys,
			fmt.Sprintf("layer%d.attn_wq", i),
			fmt.Sprintf("layer%d.attn_wk", i),
			fmt.Sprintf("layer%d.attn_wv", i),
			fmt.Sprintf("layer%d.attn_wo", i),
			fmt.Sprintf("layer%d.mlp_fc1", i),
			fmt.Sprintf("layer%d.mlp_fc2", i),
		)
	}
	params := make([]*Value, 0)
	for _, key := range paramKeys {
		mat := stateDict[key]
		for _, row := range mat {
			params = append(params, row...)
		}
	}
	fmt.Printf("num params: %d\n", len(params))

	// GPT forward function
	gpt := func(tokenID, posID int, keys, values [][][]*Value) []*Value {
		tokEmb := stateDict["wte"][tokenID]
		posEmb := stateDict["wpe"][posID]
		x := make([]*Value, nEmbd)
		for i := range x {
			x[i] = tokEmb[i].Add(posEmb[i])
		}
		x = rmsnorm(x)

		for li := 0; li < nLayer; li++ {
			// 1) Multi-head Attention block
			xResidual := x
			x = rmsnorm(x)
			q := linear(x, stateDict[fmt.Sprintf("layer%d.attn_wq", li)])
			k := linear(x, stateDict[fmt.Sprintf("layer%d.attn_wk", li)])
			v := linear(x, stateDict[fmt.Sprintf("layer%d.attn_wv", li)])
			keys[li] = append(keys[li], k)
			values[li] = append(values[li], v)
			xAttn := make([]*Value, 0, nEmbd)
			for h := 0; h < nHead; h++ {
				hs := h * headDim
				qH := q[hs : hs+headDim]
				kH := make([][]*Value, len(keys[li]))
				vH := make([][]*Value, len(values[li]))
				for t := range keys[li] {
					kH[t] = keys[li][t][hs : hs+headDim]
					vH[t] = values[li][t][hs : hs+headDim]
				}
				scale := 1.0 / math.Sqrt(float64(headDim))
				attnLogits := make([]*Value, len(kH))
				for t := range kH {
					s := NewValue(0)
					for j := 0; j < headDim; j++ {
						s = s.Add(qH[j].Mul(kH[t][j]))
					}
					attnLogits[t] = s.MulScalar(scale)
				}
				attnWeights := softmax(attnLogits)
				headOut := make([]*Value, headDim)
				for j := 0; j < headDim; j++ {
					s := NewValue(0)
					for t := range vH {
						s = s.Add(attnWeights[t].Mul(vH[t][j]))
					}
					headOut[j] = s
				}
				xAttn = append(xAttn, headOut...)
			}
			x = linear(xAttn, stateDict[fmt.Sprintf("layer%d.attn_wo", li)])
			for i := range x {
				x[i] = x[i].Add(xResidual[i])
			}
			// 2) MLP block
			xResidual = x
			x = rmsnorm(x)
			x = linear(x, stateDict[fmt.Sprintf("layer%d.mlp_fc1", li)])
			for i := range x {
				x[i] = x[i].ReLU()
			}
			x = linear(x, stateDict[fmt.Sprintf("layer%d.mlp_fc2", li)])
			for i := range x {
				x[i] = x[i].Add(xResidual[i])
			}
		}

		logits := linear(x, stateDict["lm_head"])
		return logits
	}

	// Let there be Adam, the blessed optimizer and its buffers
	learningRate := 0.01
	beta1 := 0.85
	beta2 := 0.99
	epsAdam := 1e-8
	mBuf := make([]float64, len(params))
	vBuf := make([]float64, len(params))

	// Repeat in sequence
	numSteps := 1000
	for step := 0; step < numSteps; step++ {
		// Take single document, tokenize it
		doc := docs[step%len(docs)]
		tokens := make([]int, 0, len(doc)+2)
		tokens = append(tokens, BOS)
		for _, ch := range doc {
			tokens = append(tokens, charIndex[ch])
		}
		tokens = append(tokens, BOS)
		n := blockSize
		if len(tokens)-1 < n {
			n = len(tokens) - 1
		}

		// Forward the token sequence through the model
		keys := make([][][]*Value, nLayer)
		vals := make([][][]*Value, nLayer)
		for i := range keys {
			keys[i] = make([][]*Value, 0)
			vals[i] = make([][]*Value, 0)
		}
		losses := make([]*Value, 0, n)
		for posID := 0; posID < n; posID++ {
			tokenID := tokens[posID]
			targetID := tokens[posID+1]
			logits := gpt(tokenID, posID, keys, vals)
			probs := softmax(logits)
			lossT := probs[targetID].Log().Neg()
			losses = append(losses, lossT)
		}
		loss := sumValues(losses).MulScalar(1.0 / float64(n))

		// Backward
		loss.Backward()

		// Adam optimizer update
		lrT := learningRate * (1 - float64(step)/float64(numSteps))
		stepP1 := float64(step + 1)
		for i, p := range params {
			mBuf[i] = beta1*mBuf[i] + (1-beta1)*p.Grad
			vBuf[i] = beta2*vBuf[i] + (1-beta2)*p.Grad*p.Grad
			mHat := mBuf[i] / (1 - math.Pow(beta1, stepP1))
			vHat := vBuf[i] / (1 - math.Pow(beta2, stepP1))
			p.Data -= lrT * mHat / (math.Sqrt(vHat) + epsAdam)
			p.Grad = 0
		}

		fmt.Printf("step %4d / %4d | loss %.4f\r", step+1, numSteps, loss.Data)
	}

	// Inference: may the model babble back to us
	temperature := 0.5
	fmt.Println("\n--- inference (new, hallucinated names) ---")
	for sampleIdx := 0; sampleIdx < 20; sampleIdx++ {
		keys := make([][][]*Value, nLayer)
		vals := make([][][]*Value, nLayer)
		for i := range keys {
			keys[i] = make([][]*Value, 0)
			vals[i] = make([][]*Value, 0)
		}
		tokenID := BOS
		sample := make([]rune, 0)
		for posID := 0; posID < blockSize; posID++ {
			logits := gpt(tokenID, posID, keys, vals)
			// Apply temperature
			scaledLogits := make([]*Value, len(logits))
			for i, l := range logits {
				scaledLogits[i] = l.MulScalar(1.0 / temperature)
			}
			probs := softmax(scaledLogits)
			// Weighted random choice
			r := rng.Float64()
			cumulative := 0.0
			tokenID = 0
			for i, p := range probs {
				cumulative += p.Data
				if r <= cumulative {
					tokenID = i
					break
				}
			}
			if tokenID == BOS {
				break
			}
			sample = append(sample, uchars[tokenID])
		}
		fmt.Printf("sample %2d: %s\n", sampleIdx+1, string(sample))
	}
}
